build.sbt

```
@@ -74,7 +75,25 @@ lazy val core = crossProject(JVMPlatform, JSPlatform, NativePlatform)
   .settings(rootSettings)
   .settings(
     moduleName := "cats-tagless-core",
-    libraryDependencies += "org.typelevel" %%% "cats-core" % catsVersion
+    libraryDependencies += "org.typelevel" %%% "cats-core" % catsVersion,
+    tlVersionIntroduced := Map("3" -> "0.15.0"),
+    tlMimaPreviousVersions ++= when(scalaBinaryVersion.value.startsWith("2"))("0.14.0").toSet
+  )
+
+lazy val fs2JVM = fs2.jvm
+lazy val fs2JS = fs2.js
+lazy val fs2Native = fs2.native
+lazy val fs2 = crossProject(JVMPlatform, JSPlatform, NativePlatform)
+  .crossType(CrossType.Pure)
+  .dependsOn(core)
+  .enablePlugins(AutomateHeaderPlugin)
+  .jsSettings(commonJsSettings)
+  .nativeSettings(commonNativeSettings)
+  .settings(rootSettings)
+  .settings(
+    moduleName := "cats-tagless-fs2",
+    libraryDependencies += "co.fs2" %%% "fs2-core" % fs2Version,
+    tlVersionIntroduced := Map("2.12" -> "0.15.1", "2.13" -> "0.15.1", "3" -> "0.15.1")
```

I think it's better to be 0.16.0 - it's not really a bugfix patch release.
```suggestion
    tlVersionIntroduced := Map("2.12" -> "0.16.0", "2.13" -> "0.16.0", "3" -> "0.16.0")
```

---

I disagree—if we were past 1.x, that would be right, but since we're pre-1.x, we're in early semver mode, meaning the second position is effectively a major version bump (no implied binary compatibility), and the third is for both bug fixes and new features. 

Similarly, the `v0.15.0` release should really have been `v0.14.1`. Bumping the version to `v0.15.0` meant sbt started raising errors when evicting `v0.14.0`, even though we know they're binary compatible. I've had a bunch of build failures as a result—Scala Steward makes a PR updating cats-tagless, but a bunch of our apps use cats-tagless along with other libraries that pull cats-tagless in transitively. The Steward PRs that only update cats-tagless fail because of the sbt eviction errors unless everything is updated at once.

It's too late to do anything about `v0.15.0` now, but we can avoid future headaches by using `v0.15.1` as the next version, only going to `v0.16.0` if we need to introduce binary-incompatible changes.

---

Hmm, fair point - I wasn't aware of any libraries that pull in cats-tagless. Let's keep it `0.15` for now.
We can release 1.0 once we have are done with Scala 3 support 👍

---

Yeah, I ran with [one of your suggestions a while ago](https://github.com/typelevel/cats-tagless/pull/250) and built [Dwolla/async-utils](https://github.com/dwolla/async-utils) for safely lifting tagless-final algebras implemented in Future to cats-effect. I've also created [Dwolla/natchez-tagless](https://github.com/dwolla/natchez-tagless), which contains helpers to more easily use the AOP functionality here with Natchez tracing. (I plan to do something similar for otel4s once it's a little further along.) Our [Dwolla/fs2-aws](https://github.com/Dwolla/fs2-aws) library publishes cats-tagless typeclass instances for some of the algebras it makes available.

---

That's really cool, thanks for the references!